---
title: "Final Project"
output: pdf_document
---

```{r setup, include=FALSE}
library(TCGAbiolinks)
library(tidyr)
library(dplyr)
library(plyr)
library(ggplot2)
library(SummarizedExperiment)
library(DescTools)
library(qqman)
library(minfi)
library(purrr)
library(biomaRt)
```

## Introduction to Data and Background:

According to the Mayo Clinic, Renal Cell Carcinoma is the most common kidney cancer and the incidence of this cancer is increasing in the population [1](https://www.mayoclinic.org/diseases-conditions/kidney-cancer/symptoms-causes/syc-20352664). One of the key risk factors attributed to the onset of Renal Cell Carcinoma is familial history of the disease or transmission through genetics [1](https://www.mayoclinic.org/diseases-conditions/kidney-cancer/symptoms-causes/syc-20352664). Furthermore, there is evidence to show that kidney cancers begin when mutations form in some kidney cells' DNA which causes the cells to grow and divide very rapidly [1](). Some other studies have shown that DNA Methylation can play a role in gene expression which can in turn lead to complex diseases such as cancer[2](https://pubmed.ncbi.nlm.nih.gov/15542813/). This can specifically be seen for Renal Cell Carcinoma as well as a previous study done by Lasseigne et al showed that there is a difference in the methylation profiles between patients affected by Renal Cell Carcinoma and those that are not in several genes [3](https://pubmed.ncbi.nlm.nih.gov/25472429/). For this reason, in this analysis we will study the relationship between DNA Methylation and Renal Cell Carcinoma through the use of data gathered from the TCGA database. The data lists a set of 325 Cases and 160 Controls as well as their methylation levels. 160 of the Case samples are paired with the Control Samples meaning that the Control Sample for these 160 patients is a tissue sample that is adjacent to the tumor sample collected from the same patient. Furthermore, 165 of the Case samples are Independent samples. Studying the relationship between DNA Methylation and Renal Cell Carcinoma can help us create targeted therapies, help us to understand how this disease occurs, and help us uncover which CpG sites or regions in the genome are correlated with Renal Cell Carcinoma the most.

## Methods:

### Data Cleaning and Preprocessing:

In this analysis, we used the raw microarray data generated by Infinium Human Methylation 450K BeadChip. We followed the suggestions from previous studies [4](https://www.nature.com/articles/bjc2013496)[5](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4916993/) and used the 'minfi' package to process the raw data. We started with determining the quality of samples. We first calculated the detection p-values for every CpG site in each sample, which compares the total signal for each probe to the background signal level determined by the negative control probes. A very small p-values indicates a reliable signal; therefore, we would delete any samples with a mean detection p-value >= 0.01. Then we created a simple QCplot where good samples tend to cluster with bad samples labeled. The QCplot gave us the confidence to delete the 402nd sample from our data. Then, we normalized our data using the built in function 'preprocessFunnorm' which is suggested to be used in the case-control studies.  We then corrected the wrong predicted sex and plotted the marginal densities of beta values for the raw and normalized data. The plot shows that the normalization successfully centered the data and removed noises. Similarly, we performed quality control for probes. We first filtered out probes with any detection p-value >= 0.01. Then, we removed probes residing on the X chromosome, probes with SNPs at CpG sites and any cross-reactive probes listing in the previous study[6](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3592906/). The final number of CpG sites that we had in our analysis was 401,034. We then calculated the beta values for downstream analysis. There were two patients each submitted two tumor samples, so we took the averages to remove potential bias.


### Paired and Independent T-Tests:

To analyze the relationship between DNA Methylation and Renal Cell Carcinoma we will first perform an analysis to find differentially methylated positions (DMP) between the two groups. To do this we performed two sets of different t tests, stratifying by paired and unpaired samples. For the 159 paired Case/Control samples we will conduct a paired two sided t test for each CpG site in our data. For the other 163 Case Samples, we will conduct a two sample two sided t test comparing them with the 160 Control Samples for each CpG site in our data. Our null hypothesis for these tests is that there is no difference in methylation between cases and controls and our alternate hypothesis is that there is a significant difference in methylation between cases and controls. We will use the standard significance level of 0.05 to determine significance for both the paired and two sample t tests we are doing. We will be performing multiple comparison adjustments on this result as we are comparing hundreds of thousands of CpG sites so not adjusting for multiple comparisons could lead to large errors in our analysis. We corrected the results of the t tests using both FDR and FWER correction just to see if there was any massive difference between the two in our case. FDR would be more beneficial in our case since we have a very large number of CpG sites (over 400,000) so our power would be greatly decreased by FWER correction. 

### Plots and Figures:

The plots we chose to use to analyze the results of our t tests are manhattan plots and volcano plots. The manhattan plots plot the genomic locations in chromosome and position number on the x axis and the -log10(FDR Corrected P Value) from our t tests on the y axis. This plot was chosen so that we can pinpoint where in the genome is methylation most significantly different between cases and controls. The volcano plots plot the difference in mean normalized methylation between cases and controls on the x axis for each CpG site and on the y axis it plots the -log10(FDR Corrected P Value) from our t tests. This plot makes it easy to see which type of methylation (Hypomethylated or Hypermethylated in Cases) are most significant as well as allows us to pinpoint the most important (most significantly different) CpG sites between cases and controls.

After making the volcano plots we took the top 10 CpG sites from both the paired t tests and the independent t tests identified by the volcano plots and we looked to see if there was any patterns in genomic location, what gene the CpG sites are located in, and whether they were Hypomethylated or Hypermethylated in cases and controls. We gathered the gene location data from the Ensembl database. 

```{r echo = F, warning = F, message = F}
## Data Download, Cleaning, and Processing:

# Query 
# Query raw methylation 450K array from TCGA
query <- GDCquery(project = "TCGA-KIRC",
                      data.category = "Raw microarray data",
                      data.type = "Raw intensities", 
                      experimental.strategy = "Methylation array", 
                      legacy = TRUE,
                      file.type = ".idat",
                      platform = "Illumina Human Methylation 450")
clinical <- GDCquery_clinic(project = "TCGA-KIRC")
# Download IDAT files from TCGA
tryCatch(GDCdownload(query, method = "api", files.per.chunk = 20),
             error = function(e) GDCdownload(query, method = "client"))
# Move to current work dir
for(file in dir(".",pattern = ".idat", recursive = T)){
  TCGAbiolinks:::move(file,basename(file))
}
```


```{r echo = F, results = "hide",warning = F, message = F}
# Preprocess
#Reference: https://bioconductor.org/help/course-materials/2014/BioC2014/minfi_BioC2014.pdf
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4916993/
#https://www.nature.com/articles/bjc2013496
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5302158/
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3592906/
# Reformat to a target data frame
results <- getResults(query) %>%
  group_by(cases) %>%
  mutate(Basename=substr(file_name, 1, 17)) %>%
  dplyr::select(cases, state, sample_type, cases.submitter_id, sample.submitter_id, Basename) %>% unique() %>%
  left_join(clinical %>%
              dplyr::select(submitter_id, gender),
            by=c("cases.submitter_id"="submitter_id"))
results$sample_type[results$sample_type=="Additional - New Primary"] <- "Primary Tumor"

# rm(query)
  
# Read raw data
#setwd("/users/13038/downloads/Stat_Gen_Final_Project")
RGSets <- read.metharray.exp(targets=results)
manifest <- minfi::getManifest(RGSets)
manifest
head(getProbeInfo(manifest))
# rm(RGSets)

# Quality Control for Samples
MSets <- preprocessRaw(RGSets)
qc <- getQC(MSets)
head(qc)
plotQC(qc)
rm(MSets)

detP <- detectionP(RGSets)
detP[1:5,1:5]
keep <- colMeans(detP) < 0.01 # Remove sample with detection p-value > 0.01
table(keep) # No one is removed

# The 402nd sample seems to be bad from the QCplot, remove it
RGSets <- RGSets[,-402]
results <- results[-402,]
detP <- detP[,-402]

# Normalization (preprocessFunnorm: useful for can studies)
GRSets_norm <- preprocessFunnorm(RGSets)
plotSex(GRSets_norm)
# Some predicted sexes are wrong
GRSets_norm@colData$gender[GRSets_norm@colData$gender=="female"] <- "F"
GRSets_norm@colData$gender[GRSets_norm@colData$gender=="male"] <- "M"
GRSets_norm@colData$predictedSex <- GRSets_norm@colData$gender
plotSex(GRSets_norm)


# Beta densities for unnormalized & normalized data
par(mfrow=c(1,2))
densityPlot(RGSets, sampGroups = results$sample_type, main="Raw", legend=F)
legend("top", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)
densityPlot(getBeta(GRSets_norm), sampGroups = results$sample_type, main="Normalized", legend=F)
legend("top", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)

# Quality Control for probes
detP <- detP[match(featureNames(GRSets_norm),rownames(detP)),] # Ensure probes are in the same order
keep <- rowSums(detP < 0.01) == ncol(GRSets_norm) # Remove probes with one or more detection p-value > 0.01
table(keep)
GRSets_norm_filter <- GRSets_norm[keep,]

ann450k <-  getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
keep <- !(featureNames(GRSets_norm_filter) %in% 
            ann450k$Name[ann450k$chr %in% c("chrX", "chrY")])
table(keep) 
GRSets_norm_filter <- GRSets_norm_filter[keep,] # Remove probes on the sex chromosomes

GRSets_norm_filter <- dropLociWithSnps(GRSets_norm_filter) # Remove probes with SNPs at CpG sites

xReactiveProbes <- read.csv("/users/13038/downloads/Stat_Gen_Final_Project/data/48639-non-specific-probes-Illumina450k.csv")
keep <- !(featureNames(GRSets_norm_filter) %in% xReactiveProbes$TargetID)
table(keep)
GRSets_norm_filter <- GRSets_norm_filter[keep,] # Remove cross reactive probes

# Get M values and beta values
mVals <- getM(GRSets_norm_filter)
bVals <- getBeta(GRSets_norm_filter)
par(mfrow=c(1,2))
densityPlot(mVals, sampGroups=results$sample_type, main="M values", xlab="M values", legend=F)
legend("topleft", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)
densityPlot(bVals, sampGroups=results$sample_type, main="Beta values", xlab="Beta values", legend=F)
legend("top", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)

annotations <- getAnnotation(GRSets_norm_filter)

rm(GRSets_norm)
 
```

```{r echo = F, warning = F, message = F, results = "hide"}
## Assessing Whether Samples are Paired or Not:
multiple_inputs <- results %>%
  dplyr::group_by(sample.submitter_id) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::filter(count > 1)
which(results$sample.submitter_id==multiple_inputs$sample.submitter_id)
bVals[,27] <- (bVals[,27]+bVals[,167])/2
bVals[,92] <- (bVals[,92]+bVals[,296])/2
bVals <- bVals[,-c(167,296)]
results <- results[-c(167,296),]
colnames(bVals) <- results$sample.submitter_id

paired_samples <- results %>%
  dplyr::group_by(cases.submitter_id) %>%
  dplyr::mutate(count = n_distinct(sample_type)) %>%
  dplyr::filter(count > 1) %>% dplyr::ungroup() %>%
  dplyr::arrange(cases.submitter_id)

unpaired_controls <- results %>%
  dplyr::filter(sample_type == "Solid Tissue Normal",
         !(cases.submitter_id %in% paired_samples$cases.submitter_id)) %>%
  pull(sample.submitter_id)

unpaired_cases <- results %>%
 dplyr::filter(sample_type == "Primary Tumor",
         !(cases.submitter_id %in% paired_samples$cases.submitter_id)) %>%
  pull(sample.submitter_id)

```


```{r echo = F, warning = F, message = F}
## Paired and Independent T-Tests:
# Paired t-tests
paired_cases <- paired_samples %>%
  filter(sample_type == "Primary Tumor") %>%
  pull(sample.submitter_id)

paired_controls <- paired_samples %>%
  filter(sample_type == "Solid Tissue Normal") %>%
  pull(sample.submitter_id)

CpG_Site = rownames(bVals)

paired_t_test_res <- c() # change to a vector, so it will be faster (I guess)
for (j in 1:nrow(bVals)){
    paired_t_test_res_temp = t.test(bVals[j,paired_cases],bVals[j,paired_controls],alternative="two.sided",paired=T)$p.value
    paired_t_test_res <- c(paired_t_test_res, paired_t_test_res_temp)
}
paired_t_test_res = data.frame(p.value=paired_t_test_res)

paired_t_test_res = cbind(paired_t_test_res,CpG_Site)

paired_t_test_res_FDR = paired_t_test_res %>%
  mutate(FDR_P_Value = p.adjust(paired_t_test_res$p.value,method="fdr"),
         FWER_P_Value = p.adjust(paired_t_test_res$p.value, method="bonferroni"))

# Two-sample t-tests
ind_cases <- unpaired_cases
ind_controls <- c(unpaired_controls, paired_controls)

t_test_res <- c()
for (j in 1:nrow(bVals)){
    t_test_res_temp = t.test(bVals[j,ind_cases],bVals[j,ind_controls],alternative="two.sided")$p.value
    t_test_res <- c(t_test_res, t_test_res_temp)
}

t_test_res = data.frame(p.value=t_test_res)

t_test_res = cbind(t_test_res,CpG_Site)

t_test_res_FDR = t_test_res %>%
  mutate(FDR_P_Value = p.adjust(t_test_res$p.value,method="fdr"),
         FWER_P_Value = p.adjust(t_test_res$p.value, method="bonferroni"))

rm(t_test_res)
rm(paired_t_test_res)
```

## Results:

### Figure 1: Manhattan Plot of FDR Corrected P-Values from Paired T Test

```{r echo = F, warning = F, message = F}
chrom_info = data.frame(rowRanges(GRSets_norm_filter))%>%
  dplyr::select("seqnames","start")

chrom_info = cbind(chrom_info,GRSets_norm_filter@rowRanges@ranges@NAMES)

colnames(chrom_info)[3] = "probeID"

t_test_FDR_w_chrom = merge(paired_t_test_res_FDR,chrom_info,by.x="CpG_Site",by.y="probeID")

colnames(t_test_FDR_w_chrom) = c("CpG_Site","p.value","FDR_P_Value",
                                 "FWER_P_Value","chrom","pos")

t_test_FDR_w_chrom$FDR_P_Value = -log10(t_test_FDR_w_chrom$FDR_P_Value)

t_test_FDR_w_chrom$chrom = gsub("chr","",t_test_FDR_w_chrom$chrom)

t_test_FDR_w_chrom$chrom = recode(t_test_FDR_w_chrom$chrom,"X" = "23","Y"="24","NA"="25")

t_test_FDR_w_chrom$chrom = as.numeric(t_test_FDR_w_chrom$chrom)

plot <- t_test_FDR_w_chrom %>% 
  dplyr::group_by(chrom) %>% 
  dplyr::summarise(chr_len=max(pos)) %>% 
  dplyr::mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
  dplyr::select(-chr_len) %>%
  dplyr::left_join(t_test_FDR_w_chrom, ., by=c("chrom"="chrom")) %>%
  dplyr::arrange(chrom,pos) %>%
  dplyr::mutate(pos_cum=pos+tot)

axisdf = plot %>% dplyr::group_by(chrom) %>% dplyr::summarize(center=( max(pos_cum) + min(pos_cum) ) / 2 )


ggplot(plot, aes(x=pos_cum, y=FDR_P_Value)) +
    geom_point( aes(color=as.factor(chrom)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"),22))+
    scale_x_continuous( label = axisdf$chrom[seq(1, length(axisdf$chrom), by = 2)], breaks= axisdf$center[seq(1, length(axisdf$center), by = 2)] ) +
    scale_y_continuous(expand = c(0, 0) ) +
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
  xlab("Chromosome")+
  ylab("-log10(FDR Corrected P-value)")
```

From this manhattan plot of the paired t-test results we can see that most of the CpG sites are very significant with p values much lower than 0.05, even after FDR correction. This extremely large amount of significant CpG sites can be seen across each chromosome in the genome. More specifically we found that there were 272,139 CpG sites that were significant in the paired t test analysis and 128,895 that were not after FDR correction. This means that there is a large number of CpG sites spread out throughout the genome that have significant differential methylation between cases and controls in the TCGA-KIRC data. The most significant CpG site according to the plot is in chromosome 10. 

### Figure 2: Manhattan Plot of FWER Corrected P-Values from Paired T Test

```{r echo = F, warning = F, message = F}
chrom_info = data.frame(rowRanges(GRSets_norm_filter))%>%
  dplyr::select("seqnames","start")

chrom_info = cbind(chrom_info,GRSets_norm_filter@rowRanges@ranges@NAMES)

colnames(chrom_info)[3] = "probeID"

t_test_FDR_w_chrom = merge(paired_t_test_res_FDR,chrom_info,by.x="CpG_Site",by.y="probeID")

colnames(t_test_FDR_w_chrom) = c("CpG_Site","p.value","FDR_P_Value",
                                 "FWER_P_Value","chrom","pos")

t_test_FDR_w_chrom$FWER_P_Value = -log10(t_test_FDR_w_chrom$FWER_P_Value)

t_test_FDR_w_chrom$chrom = gsub("chr","",t_test_FDR_w_chrom$chrom)

t_test_FDR_w_chrom$chrom = recode(t_test_FDR_w_chrom$chrom,"X" = "23","Y"="24","NA"="25")

t_test_FDR_w_chrom$chrom = as.numeric(t_test_FDR_w_chrom$chrom)

plot <- t_test_FDR_w_chrom %>% 
  dplyr::group_by(chrom) %>% 
  dplyr::summarise(chr_len=max(pos)) %>% 
  dplyr::mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
  dplyr::select(-chr_len) %>%
  dplyr::left_join(t_test_FDR_w_chrom, ., by=c("chrom"="chrom")) %>%
  dplyr::arrange(chrom,pos) %>%
  dplyr::mutate(pos_cum=pos+tot)

axisdf = plot %>% dplyr::group_by(chrom) %>% dplyr::summarize(center=( max(pos_cum) + min(pos_cum) ) / 2 )


ggplot(plot, aes(x=pos_cum, y=FWER_P_Value)) +
    geom_point( aes(color=as.factor(chrom)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"),22))+
    scale_x_continuous( label = axisdf$chrom[seq(1, length(axisdf$chrom), by = 2)], breaks= axisdf$center[seq(1, length(axisdf$center), by = 2)] ) +
    scale_y_continuous(expand = c(0, 0) ) +
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
  xlab("Chromosome")+
  ylab("-log10(FWER Corrected P-value)")
```

Since there was an extremely large number of CpG sites that were signficant after FDR correction, we tried using FWER correction on the paired t test results as we thought that a lot of these CpG sites could have been deemed significant from a Type I error. From this manhattan plot of the paired t-test results we can see that most of the CpG sites are very significant with p values much lower than 0.05, even after FWER correction. This extremely large amount of significant CpG sites can be seen across each chromosome in the genome. More specifically we found that there were 152,372 CpG sites that were significant in the paired t test analysis and 248,662 that were not after FWER correction. This confirms that there is a large number of CpG sites spread out throughout the genome that have significant differential methylation between cases and controls in the TCGA-KIRC data as FWER correction is extremely strict against Type I errors so the fact that there is still such a large number of CpG sites that are significant is very promising. The most significant CpG site according to the plot is in chromosome 10.

### Figure 3: Manhattan Plot of FDR Corrected P-Values from Independent T Test

```{r echo = F, warning = F, message = F}
chrom_info = data.frame(rowRanges(GRSets_norm_filter))%>%
  dplyr::select("seqnames","start")

chrom_info = cbind(chrom_info,GRSets_norm_filter@rowRanges@ranges@NAMES)

colnames(chrom_info)[3] = "probeID"

t_test_FDR_w_chrom = merge(t_test_res_FDR,chrom_info,by.x="CpG_Site",by.y="probeID")

colnames(t_test_FDR_w_chrom) = c("CpG_Site","p.value","FDR_P_Value",
                                 "FWER_P_Value","chrom","pos")

t_test_FDR_w_chrom$FDR_P_Value = -log10(t_test_FDR_w_chrom$FDR_P_Value)

t_test_FDR_w_chrom$chrom = gsub("chr","",t_test_FDR_w_chrom$chrom)

t_test_FDR_w_chrom$chrom = recode(t_test_FDR_w_chrom$chrom,"X" = "23","Y"="24","NA"="25")

t_test_FDR_w_chrom$chrom = as.numeric(t_test_FDR_w_chrom$chrom)

plot <- t_test_FDR_w_chrom %>% 
  dplyr::group_by(chrom) %>% 
  dplyr::summarise(chr_len=max(pos)) %>% 
  dplyr::mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
  dplyr::select(-chr_len) %>%
  dplyr::left_join(t_test_FDR_w_chrom, ., by=c("chrom"="chrom")) %>%
  dplyr::arrange(chrom,pos) %>%
  dplyr::mutate(pos_cum=pos+tot)

axisdf = plot %>% dplyr::group_by(chrom) %>% dplyr::summarize(center=( max(pos_cum) + min(pos_cum) ) / 2 )

axisdf$chrom = c(seq(1,22))

ggplot(plot, aes(x=pos_cum, y=FDR_P_Value)) +
    geom_point( aes(color=as.factor(chrom)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"),22))+
    scale_x_continuous( label = axisdf$chrom[seq(1, length(axisdf$chrom), by = 2)], breaks= axisdf$center[seq(1, length(axisdf$center), by = 2)] ) +
    scale_y_continuous(expand = c(0, 0) ) +
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
  xlab("Chromosome")+
  ylab("-log10(FDR Corrected P-value)")
```

From this manhattan plot of the independent t-test results we can see that most of the CpG sites are very significant with p values much lower than 0.05, even after FDR correction. This extremely large amount of significant CpG sites can be seen across each chromosome in the genome. More specifically we found that there were 321,391 CpG sites that were significant in the independent t test analysis and 79,643 that were not after FDR correction. This means that there is a large number of CpG sites spread out throughout the genome that have significant differential methylation between cases and controls in the TCGA-KIRC data. The most significant CpG site according to the plot is in chromosome 7. Interestingly with the independent t tests we got much more of the CpG sites being listed as significant when compared to the paired t tests as well as much lower p values for many CpG sites, indicating larger significant difference in methylation between cases and controls.

### Figure 4: Manhattan Plot of FWER Corrected P-Values from Independent T Test

```{r echo = F, warning = F, message = F}
chrom_info = data.frame(rowRanges(GRSets_norm_filter))%>%
  dplyr::select("seqnames","start")

chrom_info = cbind(chrom_info,GRSets_norm_filter@rowRanges@ranges@NAMES)

colnames(chrom_info)[3] = "probeID"

t_test_FDR_w_chrom = merge(t_test_res_FDR,chrom_info,by.x="CpG_Site",by.y="probeID")

colnames(t_test_FDR_w_chrom) = c("CpG_Site","p.value","FDR_P_Value",
                                 "FWER_P_Value","chrom","pos")

t_test_FDR_w_chrom$FWER_P_Value = -log10(t_test_FDR_w_chrom$FWER_P_Value)

t_test_FDR_w_chrom$chrom = gsub("chr","",t_test_FDR_w_chrom$chrom)

t_test_FDR_w_chrom$chrom = recode(t_test_FDR_w_chrom$chrom,"X" = "23","Y"="24","NA"="25")

t_test_FDR_w_chrom$chrom = as.numeric(t_test_FDR_w_chrom$chrom)

plot <- t_test_FDR_w_chrom %>% 
  dplyr::group_by(chrom) %>% 
  dplyr::summarise(chr_len=max(pos)) %>% 
  dplyr::mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
  dplyr::select(-chr_len) %>%
  dplyr::left_join(t_test_FDR_w_chrom, ., by=c("chrom"="chrom")) %>%
  dplyr::arrange(chrom,pos) %>%
  dplyr::mutate(pos_cum=pos+tot)

axisdf = plot %>% dplyr::group_by(chrom) %>% dplyr::summarize(center=( max(pos_cum) + min(pos_cum) ) / 2 )

axisdf$chrom = c(seq(1,22))

ggplot(plot, aes(x=pos_cum, y=FWER_P_Value)) +
    geom_point( aes(color=as.factor(chrom)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"),22))+
    scale_x_continuous( label = axisdf$chrom[seq(1, length(axisdf$chrom), by = 2)], breaks= axisdf$center[seq(1, length(axisdf$center), by = 2)] ) +
    scale_y_continuous(expand = c(0, 0) ) +
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
  xlab("Chromosome")+
  ylab("-log10(FWER Corrected P-value)")
```

Since there was such a large amount of CpG sites that were significant even after FDR correction, as with the paired t tests, we wanted to see if there were still a large amount of CpG sites that are significant with FWER correction. From this manhattan plot of the independent t-test results we can see that most of the CpG sites are very significant with p values much lower than 0.05, even after FWER correction. This extremely large amount of significant CpG sites can be seen across each chromosome in the genome. More specifically we found that there were 186,342 CpG sites that were significant in the independent t test analysis and 214,692 that were not after FWER correction. This means that there is a large number of CpG sites spread out throughout the genome that have significant differential methylation between cases and controls in the TCGA-KIRC data. The most significant CpG site according to the plot is in chromosome 7. Interestingly with the independent t tests we got much more of the CpG sites being listed as significant when compared to the paired t tests as well as much lower p values for many CpG sites, indicating larger significant difference in methylation between cases and controls.

### FDR vs FWER Correction From Manhattan Plots:

Although there was a considerable reduction in the amount of CpG sites that were significant between FDR and FWER correction, around a 50% reduction for both the paired and independent t tests we will choose to use the FDR corrected results for further analyses. This is because a large proportion of the CpG sites still remained significant after FWER correction (almost 50% for both the paired and independent t tests) and the FWER correction is very strict when it comes to controlling for Type I error. Since there is still such a large number of CpG sites that are significant even after such a stringent multiple testing criteria was applied, we feel that we would only be losing power in our analysis if we use FWER correction instead of FDR correction. 

### Figure 5: Volcano Plot for Paired Samples

```{r echo = F, warning = F, message = F}
bVals_t = data.frame(t(bVals))

diff_mean_paired_Case = map_dfr(bVals_t[paired_cases,],mean,na.rm=T)
diff_mean_paired_Control = map_dfr(bVals_t[paired_controls,],mean,na.rm=T)
diff_mean_paired = diff_mean_paired_Case-diff_mean_paired_Control
diff_mean_paired = pivot_longer(diff_mean_paired,1:ncol(diff_mean_paired),values_to = "diff_mean", names_to = "CpG_Site")
diff_mean_paired = merge(diff_mean_paired,paired_t_test_res_FDR,by="CpG_Site")
diff_mean_paired$diffexpressed <- "Not Significant"
diff_mean_paired$diffexpressed[diff_mean_paired$diff_mean > 0.25 & diff_mean_paired$FDR_P_Value < 10^-5] <- "Hypermethylated"
diff_mean_paired$diffexpressed[diff_mean_paired$diff_mean < 0.25 & diff_mean_paired$FDR_P_Value < 10^-5] <- "Hypomethylated"
diff_mean_paired$diffexpressed[diff_mean_paired$diff_mean <= 0.25 & diff_mean_paired$diff_mean >= -0.25&diff_mean_paired$FDR_P_Value < 10^-5] <- "No Difference In Methylation"

ggplot(diff_mean_paired, aes(x=diff_mean, y=-log10(FDR_P_Value),col=diffexpressed)) +
    geom_point(alpha=0.8, size=1.3)+
    theme_bw() +
  scale_color_manual(values=c("black", "blue", "pink","red"))+
    theme(legend.position="top")+
  guides(fill=guide_legend(title="Methylation (Case vs Controls)"))+
  geom_hline(yintercept = 5, color = "darkblue")+
  geom_vline(xintercept = 0.25,color="darkred")+
  geom_vline(xintercept = -0.25,color="darkred")+
  xlab("Difference in Mean Methylation between Case and Controls")+
  ylab("-log10(FDR Corrected P-value)")+
  ggtitle("Volcano Plot for Paired Case/Control Samples")
```

From this Volcano Plot for the paired Case/Control samples we can see that much more CpG sites that have a difference in mean methylation values between Cases and Controls are Hypomethylated in Cases when compared to Controls. There is also a lot of CpG sites that are classified as signficant in the plot that do not show much of a difference in mean methylation (is between plus or minus .25 in value) between Cases and Controls. 

### Figure 6: Volcano Plot for Independent Samples

```{r echo = F, warning = F, message = F}
diff_mean_Idep_Case = map_dfr(bVals_t[ind_cases,],mean,na.rm=T)
diff_mean_Idep_Control = map_dfr(bVals_t[ind_controls,],mean,na.rm=T)
diff_mean_Idep = diff_mean_Idep_Case-diff_mean_Idep_Control
diff_mean_Idep = pivot_longer(diff_mean_Idep,1:ncol(diff_mean_Idep),values_to = "diff_mean", names_to = "CpG_Site")
diff_mean_Idep = merge(diff_mean_Idep,t_test_res_FDR,by="CpG_Site")
diff_mean_Idep$diffexpressed <- "Not Significant"
diff_mean_Idep$diffexpressed[diff_mean_Idep$diff_mean > 0.25 & diff_mean_Idep$FDR_P_Value < 10^-5] <- "Hypermethylated"
diff_mean_Idep$diffexpressed[diff_mean_Idep$diff_mean < 0.25 & diff_mean_Idep$FDR_P_Value < 10^-5] <- "Hypomethylated"
diff_mean_Idep$diffexpressed[diff_mean_Idep$diff_mean <= 0.25 & diff_mean_Idep$diff_mean >= -0.25&diff_mean_Idep$FDR_P_Value < 10^-5] <- "No Difference In Methylation"

ggplot(diff_mean_Idep, aes(x=diff_mean, y=-log10(FDR_P_Value),col=diffexpressed)) +
    geom_point(alpha=0.8, size=1.3)+
    theme_bw() +
  scale_color_manual(values=c("black", "blue", "pink","red"))+
    theme(legend.position="top")+
  guides(fill=guide_legend(title="Methylation (Case vs Controls)"))+
  geom_hline(yintercept = 5, color = "darkblue")+
  geom_vline(xintercept = 0.25,color="darkred")+
  geom_vline(xintercept = -0.25,color="darkred")+
  xlab("Difference in Mean Methylation between Case and Controls")+
  ylab("-log10(FDR Corrected P-value)")+
  ggtitle("Volcano Plot for Independent Case/Control Samples")
```

From this Volcano Plot for the independent Case/Control samples we can see that much more CpG sites that have a difference in mean methylation values between Cases and Controls are Hypomethylated in Cases when compared to Controls. As with the paired samples volcano plot, there is also a lot of CpG sites that are classified as signficant in the plot that do not show much of a difference in mean methylation (is between plus or minus .25 in value) between Cases and Controls. 

### Top 10 CpG Sites Identified By Volcano Plots Analysis:

```{r echo = F, warning = F, message = F}
chrom_info = data.frame(CpG_Site = GRSets_norm_filter@rowRanges@ranges@NAMES,rowRanges(GRSets_norm_filter))

top_10_paired = arrange(diff_mean_paired,FDR_P_Value)

top_10_paired = filter(top_10_paired,!diffexpressed=="No Difference In Methylation"&!diffexpressed=="Not Significant")

top_10_paired = head(top_10_paired,10)

top_10_paired = merge(top_10_paired,chrom_info,by="CpG_Site")

top_10_paired$seqnames = gsub("chr","",top_10_paired$seqnames)

mart_export <- read.delim("C:/Users/13038/Downloads/Stat_Gen_Final_Project/mart_export.txt")

all.genes = data.frame()

for (i in 1:nrow(top_10_paired)){
  for (j in 1:nrow(mart_export)){

if (top_10_paired$seqnames[i]==mart_export$Chromosome.scaffold.name[j]&
    top_10_paired$start[i]>=mart_export$Gene.start..bp.[j]&
    top_10_paired$start[i]<=mart_export$Gene.end..bp.[j]){
  all.genes.tmp = cbind(top_10_paired$CpG_Site[i],mart_export$HGNC.symbol[j],mart_export$Gene.stable.ID[j])
}
  }
all.genes = rbind(all.genes,all.genes.tmp)
}

top_10_idep = arrange(diff_mean_Idep,FDR_P_Value)

top_10_idep = filter(top_10_idep,!diffexpressed=="No Difference In Methylation"&!diffexpressed=="Not Significant")

top_10_idep = head(top_10_idep,10)

top_10_idep = merge(top_10_idep,chrom_info,by="CpG_Site")

top_10_idep$seqnames = gsub("chr","",top_10_idep$seqnames)

all.genes.idep = data.frame()

for (i in 1:nrow(top_10_idep)){
  for (j in 1:nrow(mart_export)){

if (top_10_idep$seqnames[i]==mart_export$Chromosome.scaffold.name[j]&
    top_10_idep$start[i]>=mart_export$Gene.start..bp.[j]&
    top_10_idep$start[i]<=mart_export$Gene.end..bp.[j]){
  all.genes.tmp = cbind(top_10_idep$CpG_Site[i],mart_export$HGNC.symbol[j],mart_export$Gene.stable.ID[j])
}
  }
all.genes.idep = rbind(all.genes.idep,all.genes.tmp)
}

paired_gene_table = merge(all.genes,top_10_paired,by.x = "V1",by.y="CpG_Site")

paired_gene_table=paired_gene_table[-10,]

paired_gene_table = dplyr::select(paired_gene_table,c(seq(1,4),6,8))

paired_gene_table$V2[2] = "AC073332.1"

paired_gene_table = rbind(paired_gene_table,
                          data.frame(V1 = top_10_paired$CpG_Site[10],
                                     V2 = "N/A",
                                     V3 = "N/A",
                              diff_mean = top_10_paired$diff_mean[10],
                          FDR_P_Value = top_10_paired$FDR_P_Value[10],
                          diffexpressed = top_10_paired$diffexpressed[10]))

colnames(paired_gene_table) = c("CpG Site","Gene ID", "Ensembl Gene ID", "Difference in Mean Methylation between Cases and Controls","FDR-Corrected P Value","Methylation Status (Case vs Controls)")

paired_gene_table%>%knitr::kable(caption="Top 10 CpG Sites Prioritized from Paired T Test Volcano Plot and Genes they are located in",digits = 5)

idep_gene_table = merge(all.genes.idep,top_10_idep,by.x = "V1",by.y="CpG_Site")

idep_gene_table=idep_gene_table[c(1,2,3,6,8),]

idep_gene_table = dplyr::select(idep_gene_table,c(seq(1,4),6,8))

idep_gene_table$V2[1] = "AC073332.1"

idep_gene_table = rbind(idep_gene_table,
                          data.frame(V1 = top_10_idep$CpG_Site[c(1,5,6,8,10)],
                                     V2 = "N/A",
                                     V3 = "N/A",
                              diff_mean = top_10_idep$diff_mean[c(1,5,6,8,10)],
                          FDR_P_Value = top_10_idep$FDR_P_Value[c(1,5,6,8,10)],
                          diffexpressed = top_10_idep$diffexpressed[c(1,5,6,8,10)]))

colnames(idep_gene_table) = c("CpG Site","Gene ID", "Ensembl Gene ID", "Difference in Mean Methylation between Cases and Controls","FDR-Corrected P Value","Methylation Status (Case vs Controls)")

idep_gene_table%>%knitr::kable(caption="Top 10 CpG Sites Prioritized from Independent T Test Volcano Plot and Genes they are located in",digits = 5)
```

### Overlap between Paired and Independent Tests:
```{r echo = F, warning = F, message = F}
# FDR
sig_paired_FDR <- paired_t_test_res_FDR$CpG_Site[paired_t_test_res_FDR$FDR_P_Value < 0.05]
sig_ind_FDR <- t_test_res_FDR$CpG_Site[t_test_res_FDR$FDR_P_Value < 0.05]
overlap_FDR <- intersect(sig_paired_FDR, sig_ind_FDR)
nonoverlap_FDR <- c(sig_paired_FDR[!(sig_paired_FDR %in% overlap_FDR)], sig_ind_FDR[!(sig_ind_FDR %in% overlap_FDR)])

# FWER
sig_paired_FWER <- paired_t_test_res_FDR$CpG_Site[paired_t_test_res_FDR$FWER_P_Value < 0.05]
sig_ind_FWER <- t_test_res_FDR$CpG_Site[t_test_res_FDR$FWER_P_Value < 0.05]
overlap_FWER <- intersect(sig_paired_FWER, sig_ind_FWER)
nonoverlap_FWER <- c(sig_paired_FWER[!(sig_paired_FWER %in% overlap_FWER)], sig_ind_FWER[!(sig_ind_FWER %in% overlap_FWER)])
```


## Discussion and Conclusions:

## References:
